File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:
    bool predictTheWinner(vector<int>& nums) {
        int n = nums.size();
        if(n % 2 == 0)
        {
            return true;
        }
        return solve(nums, 0, n - 1) >= 0;
    }
private:
    int solve(vector<int>& nums, int i, int j)
    {
        if(i == j)
        {
            return nums[i];
        }
        return max(nums[i] - solve(nums, i + 1, j), nums[j] - solve(nums, i, j - 1));
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def predictTheWinner(self, nums: List[int]) -> bool:
        if len(nums) % 2 == 0:
            return True
        def solve(i: int, j: int) -> int:
            if i == j:
                return nums[i]
            return max(nums[i] - solve(i + 1, j), nums[j] - solve(i, j - 1))
        return solve(0, len(nums) - 1) >= 0